### ISSUE TITLE
Converting to and from VarULE

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#1179]
## Converting From VarULE

From VarULE is fairly easy.  `From` works fine when the target type allocates memory.  If we want to support cases where the target type borrows from the VarULE, `FromVarULE` can work ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1bf93bd34bff6a0a96039a3ba9cc2089)):

```rust
trait FromVarULE<'a, U: ?Sized>: 'a {
    fn from_var_ule(ule: &'a U) -> Self;
}

impl FromVarULE<'_, str> for String {
    fn from_var_ule(ule: &str) -> Self {
        ule.to_string()
    }
}

impl<'a> FromVarULE<'a, str> for Cow<'a, str> {
    fn from_var_ule(ule: &'a str) -> Self {
        Cow::Borrowed(ule)
    }
}
```

## Converting To VarULE

The other direction is more complicated because VarULE is unsized.

### Box-based solutions

A simple trait would be something like the following, but we should not consider it because it requires allocation ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=33033d47ac2f60148d98a678120c11b5)):

```rust
trait IntoBoxedVarULE<U: ?Sized> {
    fn into_var_ule(&self) -> Box<U>;
}

impl IntoBoxedVarULE<str> for String {
    fn into_var_ule(&self) -> Box<str> {
        self.clone().into_boxed_str()
    }
}

impl IntoBoxedVarULE<str> for Cow<'_, str> {
    fn into_var_ule(&self) -> Box<str> {
        self.to_string().into_boxed_str()
    }
}
```

An interesting option would be a `Box`-based solution that uses a custom allocator such that it does not need to actually allocate memory.  A trait definition such as the following would be cool, but I can't get this to compile because `MaybeUninit` requires a `Sized` type (which seems silly to me, since the main reason you allocate in Rust is for unsized things):

```rust
// error[E0277]: the size for values of type `U` cannot be known at compilation time
trait IntoAllocBoxedVarULE<U: ?Sized> {
    fn into_var_ule<A: Allocator>(
        &self,
        allocf: impl FnOnce(usize) -> Box<MaybeUninit<U>, A>
    ) -> Box<U, A>;
}
```

A workaround would be to pass an uninitialized buffer into the function.  This results in a safe trait (I think), although implementing it requires unsafe code and unstable features ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=105ba2d7665e2efb15d4fb19e11650a7)):

```rust
trait IntoAllocBufferVarULE<U: ?Sized> {
    fn into_var_ule<A: Allocator>(
        &self,
        allocf: impl FnOnce(usize) -> Box<[MaybeUninit<u8>], A>
    ) -> Box<U, A>;
}
```

### Buffer-based solutions

@Manishearth proposed the following in #1173:

```rust
pub unsafe trait EncodeAsVarULE {
    type VarULE: VarULE + ?Sized;
    fn encode_var_ule<R>(&self, cb: impl FnOnce(&[&[u8]]) -> R) -> R;
}

unsafe impl EncodeAsVarULE for String {
    type VarULE = str;
    fn encode_var_ule<R>(&self, cb: impl FnOnce(&[&[u8]]) -> R) -> R {
        cb(&[self.as_bytes()])
    }
}
```

The advantage of this type of approach, which I'll call a "buffer-based approach" since it returns a `[u8]` instead of a `U`, is that it is easy to reason about and doesn't require unstable features or additional memory allocations.  The disadvantage is that it requires the trait to be unsafe.

An issue with the above trait is that it requires creating the `&[&[u8]]`, which is easy if the outer slice has only a single element, but may require allocating if multiple slices are required (such as a `Vec<String>`).  An alternative would be something such as:

```rust
trait AppendableBuffer {
    fn push_bytes(&mut self, bytes: &[u8]);
}

unsafe trait AppendAsVarULE<U: ?Sized> {
    fn encode_var_ule<R, A: AppendableBuffer>(
        &self,
        appendable: &mut A
    ) -> R;
}
```

The above solution could also use `std::io::Write`, but that trait requires the `std` feature.

We could also pre-allocate the whole buffer:

```rust
unsafe trait WriteToBufferAsVarULE<U: ?Sized> {
    fn encode_var_ule<'a, E>(
        &self,
        get_buffer: impl FnOnce(usize) -> &'a mut [u8]
    ) -> Result<&'a U, E>;
}
```

An advantage of `WriteToBufferAsVarULE` is that the safety constraint is a bit simpler: the only requirement is that the pointer returned by `encode_var_ule` is equal in in location, size, and alignment to the pointer returned by `get_buffer`.  Validating that the buffer is a valid `VarULE` is done internally in the function.

CC @zbraniecki

**Comment by Manishearth** [comment]
This will be explained a bit better in the EncodeAsVarULE docs once I get to them :)

**Comment by sffc** [comment]
What is the use case you had in mind where the slice-of-slices has more than one element?

**Comment by sffc** [comment]
Okay.  So this is basically a variadic function, and [using a slice seems to be Rust convention for that](https://stackoverflow.com/q/28951503).  Works for me

**Comment by Manishearth** [comment]
Yeah should have opened with that :smile: . I actually considered trying to make this work with generic tuples instead (because then you could just feed it ULE types directly) but then you need an extra trait and you have to manually implement it on all kinds of tuples.

**Comment by Manishearth** [comment]
Hmmm. So an important thing to note is that due to the use of generics it's getting inlined _anyway_, and most likely optimized; using a defaulted function can be a helpful way to hint (and reduces duplication) but I'd rather not require people to implement it manually.

**Comment by sffc** [comment]
Discussion from today: we solve almost all of our known problems with

```rust
unsafe trait EncodeAsVarULE<U: ?Sized> {
    fn encode_var_ule<'a, E>(
        &self,
        get_buffer: impl FnOnce(usize) -> &'a mut [u8]
    ) -> Result<&'a U, E>;

    fn encode_len(&self) -> usize;

    fn encode_write<'a, E>(&self, &'a mut [u8]) -> Result<&'a U, E>;
}
```

where `encode_len` and `encode_write` are auto-implemented to call `encode_var_ule`, but they can be overridden.  Meanwhile, clients such as VarZeroVec::insert() will only call `encode_len` and `encode_write` and never call `encode_var_ule` except indirectly via those methods.
